### PR TITLE
[Buttons] Add fallback behavior to setBorderColor API.

### LINKED ISSUE #3409: [Buttons] Unset `borderColorForState:` reverts to 0, not .normal color value
### Overview
MDCButton's `borderColorForState:` will return `nil` if a state is not explicitly set. This may be intentional, but does not match the behavior of `titleForState:`, `titleColorForState:`, `titleShadowColorForState:` nor `attributedTitleForState:`.  I think it will be more expected for clients if the border width defaults to the value for the `.normal` state to match the related properties.

### Reproduction steps

```objectivec
- (void)testButtonBorderColorForStateFallbackBehavior {
  // Given
  MDCButton *button = [[MDCButton alloc] init];

  // When
  [button setBorderColor:UIColor.magentaColor forState:UIControlStateNormal];

  // Then
  UIControlState maximalControlState = UIControlStateNormal | UIControlStateSelected |
      UIControlStateDisabled | UIControlStateHighlighted;
  for (NSUInteger state = 0; state <= maximalControlState; ++state) {
      XCTAssertEqualObjects([button borderColorForState:state], UIColor.magentaColor);
   }
}
```

### Version number
54.8.0

---
#### Internal data
- Associated internal bug: [b/125311810](http://b/125311810)

### PR DESCRIPTION
**Pull request by fluxminds**
closes https://github.com/material-components/material-components-ios/issues/3409

### CODE REVIEW

File: components/Buttons/src/MDCButton.m [hunk 1]
```diff
@@ -667,7 +667,7 @@ - (void)setElevation:(CGFloat)elevation forState:(UIControlState)state {
 #pragma mark - Border Color
 
 - (UIColor *)borderColorForState:(UIControlState)state {
-  return _borderColors[@(state)];
+  return _borderColors[@(state)] ?: _borderColors[@(UIControlStateNormal)];
 }
 
 - (void)setBorderColor:(UIColor *)borderColor forState:(UIControlState)state {
```
Comment (romoore) [inline components/Buttons/src/MDCButton.m:676]: Where are the changes to `setBorderColor`? Please see related methods like `setBorderWidth:forState:` and https://github.com/material-components/material-components-ios/pull/5919 for examples and context about how these properties should work.  There are complexities around `UIControlStateHighlighted` and `UIControlStateDisabled` that need to be accounted for.

**Review by romoore — DISMISSED** [review]
Please correct the fallback behavior to match that of `titleColorForState:` and other similar APIs.

File: components/Buttons/src/MDCButton.m (hunk 1 above)
Comment (fluxminds) [reply components/Buttons/src/MDCButton.m:676]: Addressed in 9edc2664fa337b4dcfc963d30246ee5c44492768

File: components/Buttons/src/MDCButton.m [hunk 2]
```diff
@@ -667,10 +667,16 @@ - (void)setElevation:(CGFloat)elevation forState:(UIControlState)state {
...
 
 - (UIColor *)borderColorForState:(UIControlState)state {
+  if ((state & UIControlStateHighlighted) == UIControlStateHighlighted) {
+    state = state & ~UIControlStateDisabled;
+  }
   return _borderColors[@(state)] ?: _borderColors[@(UIControlStateNormal)];
 }
 
 - (void)setBorderColor:(UIColor *)borderColor forState:(UIControlState)state {
+  if ((state & UIControlStateHighlighted) == UIControlStateHighlighted) {
```
Comment (romoore) [inline components/Buttons/src/MDCButton.m:679]: Looks like this is still missing some logic.  Please see #5919 for the logic required to match behavior.

```
// Only update the backing dictionary if:
// 1. The `state` argument is the same as the "storage" state, OR
// 2. There is already a value in the "storage" state.
```

**Review by romoore — DISMISSED** [review]
The behavior does not match the reference PR or UIButton APIs. Please be sure to match the behavior.  If necessary, consider using the tests outlined in #5919 that I used to validate other properties.

File: components/Buttons/src/MDCButton.m (hunk 2 above)
Comment (fluxminds) [reply components/Buttons/src/MDCButton.m:679]: I am curious about the logic here:

> 
```
  // Only update the backing dictionary if:
  // 1. The `state` argument is the same as the "storage" state, OR
  // 2. There is already a value in the "storage" state.
  if (storageState == state || _backgroundColors[@(storageState)] != nil) {
    _backgroundColors[@(storageState)] = backgroundColor;
    [self updateAlphaAndBackgroundColorAnimated:NO];
  }
```
In this case, the `_backgroundColors` is not set when `(storageState != state && _backgroundColors[@(storageState)] == nil`, which only happens when state = highlight & disable. If we perform `state = state & ~UIControlStateDisabled;` `highlight & disable` also never gets set, which should behave the same unless `_backgroundColor[highlight & disable]` is accidentally set by some other method.

Did I understand it correctly?

**Comment by fluxminds** [comment]
@romoore For testing, I added one according to the PR you mentioned. What are the other properties you tested?

**Comment by romoore** [comment]
@wenyuzhang666 Let's discuss offline. There seems to be some confusion about how the property should behave. Perhaps the Issue needs greater clarification.

File: components/Buttons/src/MDCButton.m (hunk 2 above)
Comment (romoore) [reply components/Buttons/src/MDCButton.m:679]: Not quite. The logic (simplified) is this:

1.  Attempt to set `.highlighted`
    1.  Always succeeds
1.  Attempt to set `.disabled`
    1.  Always succeeds
1.  Attempt to set `.highlighted | .disabled`
    1.  If `.highlighted` is set, override the value
    2.  If `.highlighted` is unset, do nothing
1.  Attempt to set `.selected | .highlighted | .disabled`
    1.  If `.selected | .highlighted` is set, override the value
    2.  If `.selected | .highlighted` is unset, do nothing

Comment (fluxminds) [reply components/Buttons/src/MDCButton.m:679]: I see. Thanks for the clarification:) The logic is interesting.

Addressed in 8e10afe46d3716fac6ca42aba4812b11d673bba3.

**Comment by fluxminds** [comment]
Added another backward test. Though it doesn't cover all possible state dependencies, it is helpful since our focus is the relationship between `Highlight` and  `Highlight & Disable`.

File: components/Buttons/src/MDCButton.m [hunk 3]
```diff
@@ -667,13 +667,25 @@ - (void)setElevation:(CGFloat)elevation forState:(UIControlState)state {
 #pragma mark - Border Color
 
 - (UIColor *)borderColorForState:(UIControlState)state {
-  return _borderColors[@(state)];
+  if ((state & UIControlStateHighlighted) == UIControlStateHighlighted) {
```
Comment (romoore) [inline components/Buttons/src/MDCButton.m:670]: Tip: Please document this code. You found it challenging to work with and others likely will as well.

```
 // If the `.highlighted` flag is set, turn off the `.disabled` flag
```

**Review by romoore — APPROVED** [review]

File: components/Buttons/src/MDCButton.m (hunk 3 above)
Comment (fluxminds) [reply components/Buttons/src/MDCButton.m:670]: Addressed in 10380b5002dd5accb44dbd0c06d99501a273be70

**Pull request merged**